handlers/infobip/infobip.go

```diff
@@ -1,3 +1,248 @@
 package infobip
 
-/* no logs */
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/nyaruka/courier/utils"
+	"github.com/nyaruka/gocommon/urns"
+	"github.com/pkg/errors"
+)
+
+var sendURL = "https://api.infobip.com/sms/1/text/advanced"
+
+func init() {
+	courier.RegisterHandler(NewHandler())
+}
+
+type handler struct {
+	handlers.BaseHandler
+}
+
+// NewHandler returns a new Infobip handler
+func NewHandler() courier.ChannelHandler {
+	return &handler{handlers.NewBaseHandler(courier.ChannelType("IB"), "Infobip")}
+}
+
+// Initialize is called by the engine once everything is loaded
+func (h *handler) Initialize(s courier.Server) error {
+	h.SetServer(s)
+	err := s.AddReceiveMsgRoute(h, "POST", "receive", h.ReceiveMessage)
+	if err != nil {
+		return err
+	}
+	return s.AddUpdateStatusRoute(h, "POST", "delivered", h.StatusMessage)
+}
+
+// StatusMessage is our HTTP handler function for status updates
+func (h *handler) StatusMessage(channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.MsgStatus, error) {
+	ibStatusEnvelop := &ibStatusEnvelop{}
+	err := handlers.DecodeAndValidateJSON(ibStatusEnvelop, r)
+	if err != nil {
+		return nil, courier.WriteError(w, r, err)
+	}
+
+	msgStatus, found := infobipStatusMapping[ibStatusEnvelop.Results[0].Status.GroupName]
+	if !found {
+		return nil, courier.WriteError(w, r, fmt.Errorf("unknown status '%s', must be one of PENDING, DELIVERED, EXPIRED, REJECTED or UNDELIVERABLE", ibStatusEnvelop.Results[0].Status.GroupName))
+	}
+
+	// write our status
+	status := h.Backend().NewMsgStatusForID(channel, courier.NewMsgID(ibStatusEnvelop.Results[0].MessageID), msgStatus)
+	err = h.Backend().WriteMsgStatus(status)
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.MsgStatus{status}, courier.WriteStatusSuccess(w, r, []courier.MsgStatus{status})
+}
+
+var infobipStatusMapping = map[string]courier.MsgStatusValue{
+	"PENDING":       courier.MsgSent,
+	"EXPIRED":       courier.MsgSent,
+	"DELIVERED":     courier.MsgDelivered,
+	"REJECTED":      courier.MsgFailed,
+	"UNDELIVERABLE": courier.MsgFailed,
+}
+
+type ibStatusEnvelop struct {
```

Comment:
`envelope` not `envelop`

Reply:
It'd be nice to have an example payload here as a comment just as we do for received messages below.